internal/pbutil/shallow_equal.go

```diff
@@ -0,0 +1,170 @@
+//     both use the same names for the same types, despite declaring them in
+//     inverted order.
+//   - Instances of C can be equal with instances of D or E for the same
+//     reason that instances of A and instances of B can be equal.
+//   - Instances of D can also be equal with instances of E because they
+//     declare the same type of fields in the same order, despite using
+//     different names for the fields.
+//   - However, instances of D cannot be equal with instances of B because
+//     they use different names and in a different order.
+//   - Instances of F cannot be equal with instances of any other type because
+//     F only has one field whereas every other type has two.
+//   - Instances of G cannot be equal with instances of any other type because
+//     G uses a field type that none of the other types do (i.e., `float32` vs
+//     the `int32` in everything else.) Even though G named its first field
+//     `MyInt`, it is not be considered equal to an `int32` (since this
+//     function uses `Value.Equal()` to compare the field values.)
+//
+// ```
+//
+//	type A struct {
+//		int32
+//		string
+//	}
+//
+//	type B struct {
+//		MyInt int32
+//		MyString string
+//	}
+//
+//	type C struct {
+//		string
+//		int32
+//	}
+//
+//	type D struct {
+//		MyString string
+//		MyInt int32
+//	}
+//
+//	type E struct {
+//		SomeString string
+//		SomeInt int32
+//	}
+//
+//	type F struct {
+//		MyInt int32
+//	}
+//
+//	type G struct {
+//		MyInt float32	// intentionally lying about the type with this name
+//		MyString string
+//	}
+//
+// ```
+//
+// Otherwise, if they're some other type that is `.Comparable()`, it just
+// does `Value.Equal()` and hopes for the best.
+//
+// If none of the above apply, it returns false.
+//
+// Superceding any of the above, if an error occurs, false will be returned
+// along with the error.
+//
+// Note that this method is different from `reflect.Equal` in that it does not
+// care if the types of the arguments are the same, only their `Kind`s.
+func ShallowEqual(a, b any) (equal bool, err error) {
+	aVal, bVal := reflect.Indirect(reflect.ValueOf(a)), reflect.Indirect(reflect.ValueOf(b))
+	if aVal.Kind() == reflect.Struct && bVal.Kind() == reflect.Struct {
+		aFields, bFields := reflect.VisibleFields(aVal.Type()), reflect.VisibleFields(bVal.Type())
+		aComparableFields, bComparableFields := Filter(aFields, fieldTypeIsComparable), Filter(bFields, fieldTypeIsComparable)
+		if len(aComparableFields) != len(bComparableFields) {
+			return
+		}
+		// assume equal until proven otherwise
+		equal = true
+		// check by index first, since it's easier
+		for i, aField := range aComparableFields {
+			aFieldVal, aErr := aVal.FieldByIndexErr(aField.Index)
+			bFieldVal, bErr := bVal.FieldByIndexErr(bComparableFields[i].Index)
+			err = errors.Join(err, aErr, bErr)
+			equal = equal && (err == nil) && aFieldVal.Equal(bFieldVal)
```

Comment:
Should the loop break as soon as `equal==false`? There is no reason to loop till the last field.

Reply:
Good point, I was so focused on reducing line count and making the code look more elegant that I let it be less performant. I'll have it break the loop

Reply:
Actually, no, I take that back. I remember now. I was letting it collect errors over each iteration. I do actually want to let this keep running until the end. But in the next loop I will still insert a break.